modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/operation/ComponentMessageProcessor.java

```diff
@@ -437,18 +441,62 @@ private void onEventSynchronous(CoreEvent event, ExecutorCallback executorCallba
         prepareAndExecuteOperation(event, () -> callback, ctx);
       };
 
-      operationExecutionFunction.execute(resolutionResult, event, executorCallback);
+      operationExecutionFunction.execute(resolutionResult, event,
+                                         computeOperationReturnExecutionCallback(event, executorCallback));
     } catch (Throwable t) {
       executorCallback.error(unwrap(t));
     }
   }
 
+  private ExecutorCallback computeOperationReturnExecutionCallback(CoreEvent event, ExecutorCallback delegateCallback) {
+    return new ExecutorCallback() {
+
+      @Override
+      public void complete(Object o) {
+        ExecutionContextAdapter operationContext = null;
+        CoreEvent originalEvent = null;
+        try {
+          if (returnDelegate instanceof ValueReturnDelegate) {
+            delegateCallback.complete(o);
+          } else {
+            OperationExecutionParams operationExecutionParams =
+                from(event).getOperationExecutionParams(getLocation(), event.getContext().getId());
+
+            if (operationExecutionParams != null) {
+              operationContext =
+                  operationExecutionParams.getExecutionContextAdapter();
+              originalEvent = operationContext.getEvent();
+              operationContext.changeEvent(event);
+            } else {
+              operationContext = createExecutionContext(event);
+            }
+
+            delegateCallback.complete(returnDelegate.asReturnValue(o, operationContext));
+          }
+        } catch (MuleException e) {
+          delegateCallback.error(e);
+        } catch (Throwable t) {
+          delegateCallback.error(unwrap(t));
+        } finally {
+          if (operationContext != null && originalEvent != null) {
+            operationContext.changeEvent(originalEvent);
+          }
+        }
+      }
+
+      @Override
+      public void error(Throwable t) {
+        delegateCallback.error(unwrap(t));
+      }
+    };
+  }
+
   private ExecutorCallback mapped(ExecutorCallback callback, ExecutionContextAdapter<T> operationContext) {
     return new ExecutorCallback() {
 
       @Override
       public void complete(Object value) {
-        callback.complete(returnDelegate.asReturnValue(value, operationContext));
+        callback.complete(valueReturnDelegate.asReturnValue(value, operationContext));
```

Comment:
isn't this the other way around? if there are no policies, this is the callback wrapper that is used, and you're wrongfully changing the return delefate in this case